pkg/ebpf/map_cleaner.go

```diff
@@ -107,85 +106,87 @@ func (mc *MapCleaner) Stop() {
-	for _, key := range keysToDelete {
-		err := mc.emap.Delete(key)
-		if err == nil {
-			deletedCount++
+	if len(mc.keysToDelete) > 0 {
+		count, err := mc.emap.BatchDelete(mc.keysToDelete, nil)
+		if err != nil {
+			log.Debugf("failed to delete map entries: %v", err)
+			return
 		}
+		deletedCount += count
 	}
 
-	iterationErr := entries.Err()
 	elapsed := time.Since(now)
 	log.Debugf(
-		"finished cleaning map=%s entries_checked=%d entries_deleted=%d iteration_error=%v elapsed=%s",
+		"finished cleaning map=%s entries_checked=%d entries_deleted=%d elapsed=%s",
 		mc.emap,
 		totalCount,
 		deletedCount,
-		iterationErr,
 		elapsed,
 	)
+	mc.shrinkKeysToDelete()
 }
 
-// marshalBytes converts an arbitrary value into a byte buffer.
-//
-// Returns an error if the given value isn't representable in exactly
-// length bytes.
-//
-// copied from: https://github.com/cilium/ebpf/blob/master/marshalers.go
-func marshalBytes(data interface{}, length int) (buf []byte, err error) {
-	if data == nil {
-		return nil, errors.New("can't marshal a nil value")
-	}
+func (mc *MapCleaner[K, V]) cleanWithoutBatches(nowTS int64, shouldClean func(nowTS int64, k K, v V) bool) {
+	now := time.Now()
 
-	switch value := data.(type) {
-	case encoding.BinaryMarshaler:
-		buf, err = value.MarshalBinary()
-	case string:
-		buf = []byte(value)
-	case []byte:
-		buf = value
-	case unsafe.Pointer:
-		err = errors.New("can't marshal from unsafe.Pointer")
-	default:
-		var wr bytes.Buffer
-		err = binary.Write(&wr, native.Endian, value)
-		if err != nil {
-			err = fmt.Errorf("encoding %T: %v", value, err)
+	mc.keysToDelete = mc.keysToDelete[:0]
+	totalCount, deletedCount := 0, 0
+
+	entries := mc.emap.Iterate()
+	// we resort to unsafe.Pointers because by doing so the underlying eBPF
+	// library avoids marshaling the key/value variables while traversing the map
+	for entries.Next(unsafe.Pointer(&mc.keyBatch[0]), unsafe.Pointer(&mc.valuesBatch[0])) {
+		totalCount++
+		if !shouldClean(nowTS, mc.keyBatch[0], mc.valuesBatch[0]) {
+			continue
 		}
-		buf = wr.Bytes()
+		mc.keysToDelete = append(mc.keysToDelete, mc.keyBatch[0])
 	}
-	if err != nil {
-		return nil, err
+
+	for _, key := range mc.keysToDelete {
+		err := mc.emap.Delete(unsafe.Pointer(&key))
+		if err == nil {
+			deletedCount++
+		}
 	}
 
-	if len(buf) != length {
-		return nil, fmt.Errorf("%T doesn't marshal to %d bytes", data, length)
+	elapsed := time.Since(now)
+	log.Debugf(
+		"finished cleaning map=%s entries_checked=%d entries_deleted=%d elapsed=%s",
+		mc.emap,
+		totalCount,
+		deletedCount,
+		elapsed,
+	)
+
+	mc.shrinkKeysToDelete()
+}
+
+func (mc *MapCleaner[K, V]) shrinkKeysToDelete() {
+	// If the number of keys to delete is greater than the batch size, shrinking its capacity
+	// by `mc.keysToDelete[:len(mc.keyBatch):len(mc.keyBatch)]` (similar to slices.Clip), and then resetting its length to 0
+	if len(mc.keysToDelete) > len(mc.keyBatch) {
+		mc.keysToDelete = mc.keysToDelete[:0:len(mc.keyBatch)]
```

Comment:
After thinking about this some more, this will not free up any memory. The only way it frees the backing array, is if you ever go above `len(mc.keyBatch)` again.

I'm not sure there is much value in holding onto the delete keys slice across runs. There is no predictability for the number of keys you might delete.

Reply:
changed

Reply:
What do you mean?

Reply:
You didn't make any changes in response to my above comment. Changing the length and/or capacity of a slice is not enough to release the memory used by the backing array. 

I'm suggesting to get rid of `mc.keysToDelete` and use a slice within `cleanWithBatches` and `cleanWithoutBatches`, similar to how it was done before.

Reply:
Sorry, I forgot to push the change